datafusion/sql/src/planner.rs

```diff
@@ -241,6 +186,8 @@ pub struct SqlToRel<'a, S: ContextProvider> {
     pub(crate) context_provider: &'a S,
     pub(crate) options: ParserOptions,
     pub(crate) normalizer: IdentNormalizer,
+    /// user defined planner extensions
+    pub(crate) planners: Vec<Arc<dyn UserDefinedPlanner>>,
```

Comment:
What do you think about using a single planner

```suggestion
    /// user defined planner extensions
    pub planner: Option<Arc<dyn UserDefinedPlanner>>>,
```

This would match  the basic pattern we have elsewhere (like proto extension codec) and people could implement composable planners like @lewiszlw  did in https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/composed_extension_codec.rs

Reply:
I'm not sure about this

1. If we only accept single planner, does that mean we need to implement all the trait functions in one struct? We can't let array function planner optional in this case. 
2. If the user expect their own planner + our builtin planner, how can they enable both? With current design, they just chain them with `with_user_defined_planner`.
```rust
query
.with_user_defined_planner(array_planner)
.with_user_defined_planner(field_access_planner)
.with_user_defined_planner(user's planner)
...
```
3. The example given https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/composed_extension_codec.rs seems to have multiple codecs too 🤔 
```rust
/// A PhysicalExtensionCodec that tries one of multiple inner codecs
/// until one works
#[derive(Debug)]
struct ComposedPhysicalExtensionCodec {
    codecs: Vec<Arc<dyn PhysicalExtensionCodec>>,
}
```